### PR TITLE
Enhance service discovery

### PR DESCRIPTION
**Pull request by wForget**
### _Why are the changes needed?_

Enhance service discovery #786

Refer to the curator-x-discovery module of the curator project for ServiceDiscovery enhancement. 

Some Interfaces:

+ `InstanceProvider`: provide instance that meet some conditions
+ `ProviderStrategy`: select an instance from the instances provided by the InstanceProvider
+ `ProviderFilter`: used in InstanceProvider to filter some instances
+ `InstanceSerializer`: serialize and deserialize instance object
+ `ServiceDiscoveryClient`: service discovery client, used to obtain the final instance

Feature：
+ support version filter
+ support tag filter
+ support random strategy
+ support strategy based on session number
+ support configure core size

TODO：

+ statistics session number
+ persistence engine version, tags and other information
+ adapt to EngineRef

Test Case: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        .creatingParentsIfNeeded()
+        .withMode(PERSISTENT)
+        .forPath(ns)
+    } catch {
+      case _: NodeExistsException => // do nothing
+      case e: KeeperException =>
+        throw new KyuubiException(s"Failed to create namespace '$ns'", e)
+    }
+    val pathPrefix = ZKPaths.makePath(
+      engine.namespace,
```
Comment (yaooqinn) [inline kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:113]: Currently, the EngineSpace(namespace) and serviceNode(engine instance) are 1 vs 1

Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:113]: Then, the ` EngineSpace` could achieve the isolation purpose for a tenant.

```scala
  /**
   * The EngineSpace used to expose itself to the KyuubiServers in `serverSpace`
   *
   * For `CONNECTION` share level:
   *   /$serverSpace_CONNECTION/$user/$sessionId
   * For `USER` share level:
   *   /$serverSpace_USER/$user[/$subDomain]
   *
   */
```

Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:113]: The issue I mentioned in #786 is that for now we can not ensure the `serviceNode` here is alive or not

Comment (wForget) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:113]: > Currently, the EngineSpace(namespace) and serviceNode(engine instance) are 1 vs 1

It doesn't seem to be one-to-one, engin namespace like: `/kyuubi_USER/test`, instance path like `/kyuubi_USER/test/serviceUri=127.0.0.1:1234;version=1.3.0;sequence=`

Comment (wForget) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:113]: > The issue I mentioned in #786 is that for now we can not ensure the serviceNode here is alive or not

Sorry, I haven't considered it. This PR is mainly to choose the more appropriate one for all engines under the same shared user.

Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: In this test, yes, it's not one-to-one. but in real-world usage, we have a distributed lock for engine creations at the server-side.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineInstance.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.ha.v2.engine
+
+case class EngineInstance(
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineInstance.scala:20]: Can we make a separate PR for this first?

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala (hunk 1 above)
Comment (wForget) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: I added a `CoreSizeStrategy` to create more instances.  When the number of engines does not reach `CoreSize`, do not return to the existing engine, but to create a new engine.

Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: I get your point. How about when the client-side uses `DBCP`? the connections will point to different engines, will this result in surprise?

**Comment by yaooqinn** [comment]
why can we just use ` curator-x-discovery`?

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+}
+
+class EngineProviderBuilder() extends InstanceProviderBuilder[EngineInstance] {
+
+  var namespace: String = _
+  var zkClient: CuratorFramework = _
+  var version: Option[String] = None
+  var tags: ArrayBuffer[String] = ArrayBuffer()
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala:71]: why this both a var and a mutable buffer?

Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala:71]: why this both a var and a mutable buffer?

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala (hunk 1 above)
Comment (wForget) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: > I get your point. How about when the client-side uses DBCP? the connections will point to different engines, will this result in surprise?

Yes, this may be problematic, but we can use a unique tag to adapt it.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/ServiceInstance.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.ha.v2
+
+case class ServiceInstance(host: String, port: Int) {
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/ServiceInstance.scala:20]: what's the relation between a ServiceInstance and EngineInstance？

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  val filters: ArrayBuffer[ProviderFilter[EngineInstance]] = ArrayBuffer()
+
+  def addFilter(filter: ProviderFilter[EngineInstance]): Unit = {
+    filters += filter
+  }
+
+  override def getInstances(): Seq[EngineInstance] = {
+    val enginePaths = zkClient.getChildren().forPath(namespace)
+    enginePaths.asScala
+      .map(getEngineInstance(_))
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala:42]: we don't try to get as much as available engine instances as we can but fail directly if any error occurs

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineInstance.scala [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    namespace: String,
+    nodeName: String,
+    host: String,
+    port: Int,
+    version: Option[String],
+    sessionNum: Int = 0,
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineInstance.scala:26]: what is the EngineInstance. sessionNum related to? a client session?

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: Does it allow only one engine for a unique tag?

Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: then the tag is a bit redundant with `[/$subDomain]`, we need to remove one of them

Comment (wForget) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: Sorry, it does not seem to be the case, it may also need to limit core size.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala (hunk 5 above)
Comment (wForget) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala:42]: Sorry, I don't quite understand this reply. The `EngineProvider.getInstances` method is to get all available engines of a shared user, and then get the most appropriate engine through `ProviderStrategy`.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala (hunk 3 above)
Comment (wForget) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala:71]: This is a problem, I will fix it.

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: Can we minimize the PR scope and do these separately?

Maybe we can start by adding a pool size for engines, and randomly pick it. then we apply picking rules to it separately in a series of PRs

Comment (yaooqinn) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: another thing is that we don't want the engine discovery to be bind to zookeeper too much in the near future, we are seeking a more light way to do this

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala (hunk 5 above)
Comment (yaooqinn) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/v2/engine/EngineProvider.scala:42]: you mention in getEngineInstance that it will throw exceptions when can not parse the data to an engine instance

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala (hunk 1 above)
Comment (wForget) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala:122]: OK, thank you. I will minimize it and implement it gradually.

**Comment by wForget** [comment]
needs to be improved, close it.
